numerics/fast_sin_cos_cycle_test.cpp

```diff
@@ -0,0 +1,150 @@
+  static constexpr int iterations_ = 1e7;
+#else
+  static constexpr int iterations_ = 1e8;
+#endif
+};
+
+// Check that we obtain the right result for special angles.
+// TODO(phl): We don't.  Improve this.
+TEST_F(FastSinCosCycleTest, SpecialValues) {
+  double sin;
+  double cos;
+  FastSinCosCycle(0.0, sin, cos);
+  EXPECT_THAT(sin, AlmostEquals(0.0, 0));
+  EXPECT_THAT(cos, AlmostEquals(1.0, 0));
+  FastSinCosCycle(0.25, sin, cos);
+  EXPECT_THAT(sin, IsNear(1.0, 1.00001));
+  EXPECT_THAT(cos, VanishesBefore(1.0, 0, 3e8));
+  FastSinCosCycle(0.5, sin, cos);
+  EXPECT_THAT(sin, AlmostEquals(0.0, 0));
+  EXPECT_THAT(cos, AlmostEquals(-1.0, 0));
+  FastSinCosCycle(0.75, sin, cos);
+  EXPECT_THAT(sin, IsNear(-1.0));
+  EXPECT_THAT(cos, VanishesBefore(1.0, 0, 3e8));
+  FastSinCosCycle(1.0, sin, cos);
+  EXPECT_THAT(sin, AlmostEquals(0.0, 0));
+  EXPECT_THAT(cos, AlmostEquals(1.0, 0));
+}
+
+// Check that, for arguments in [-1, 1], the errors are of the expected size and
+// at the expected location.
+TEST_F(FastSinCosCycleTest, Random1) {
+  std::mt19937_64 random(42);
+  std::uniform_real_distribution<> const distribution(-1.0, 1.0);
+  double max_sin_error = 0.0;
+  double max_cos_error = 0.0;
+  double max_sin_error_x = 0.0;
+  double max_cos_error_x = 0.0;
+  for (int i = 0; i < iterations_; ++i) {
+    double const x = distribution(random);
+    double sin;
+    double cos;
+    FastSinCosCycle(x, sin, cos);
+    double const sin_error = std::abs(std::sin(2 * π * x) - sin);
+    if (sin_error > max_sin_error) {
+      max_sin_error = sin_error;
+      max_sin_error_x = x;
+    }
+    double const cos_error = std::abs(std::cos(2 * π * x) - cos);
+    if (cos_error > max_cos_error) {
+      max_cos_error = cos_error;
+      max_cos_error_x = x;
+    }
+  }
+  // These numbers come from the Mathematica minimax computation.
+  EXPECT_LT(max_sin_error, 5.90e-7);
+  EXPECT_LT(max_cos_error, 5.28e-8);
+  EXPECT_THAT(max_sin_error_x, IsNear(-1.0 + 1.0 / 23.0));
+  EXPECT_THAT(max_cos_error_x, IsNear(1.0 - 1.0 / 15.0));
+}
+
+// Same as above, but in the range [-1e6, 1e6] to test argument reduction.
+TEST_F(FastSinCosCycleTest, Random1000) {
+  std::mt19937_64 random(42);
+  std::uniform_real_distribution<> const distribution(-1.0e6, 1.0e6);
+  double max_sin_error = 0.0;
+  double max_cos_error = 0.0;
+  double max_sin_error_x = 0.0;
+  double max_cos_error_x = 0.0;
+  for (int i = 0; i < iterations_; ++i) {
+    double const x = distribution(random);
+    double sin;
+    double cos;
+    FastSinCosCycle(x, sin, cos);
+    double const sin_error = std::abs(std::sin(2 * π * x) - sin);
+    if (sin_error > max_sin_error) {
+      max_sin_error = sin_error;
+      max_sin_error_x = x;
+    }
+    double const cos_error = std::abs(std::cos(2 * π * x) - cos);
+    if (cos_error > max_cos_error) {
+      max_cos_error = cos_error;
+      max_cos_error_x = x;
+    }
+  }
+  // Slightly larger errors here because of argument reduction.
```

Comment:
```c++
  // Slightly larger errors here because of argument reduction in std::sin and std::cos.
```
since our argument reduction is exact.

Perhaps it would be better to test the argument reduction by taking arbitrary double x in [-1, 1], integer n in [-1e6, 1e6], and checking the equality of the values for the arguments x+n-n and x+n, as that would check that the argument reduction is exact.

Reply:
Done.